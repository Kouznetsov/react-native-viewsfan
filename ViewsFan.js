import React, {Component} from "react";
import {
    StyleSheet,
    View,
} from "react-native"

export default class ViewsFan extends Component {

    anglesAndPositions;

    constructor(props) {
        super(props);
        this.V_OFFSET = this.props.verticalOffset;
        this.H_OFFSET = this.props.horizontalOffset;
        this.MAX_ANGLE = this.props.maxAngle;
        this.anglesAndPositions = this._genData();
    }

    _genData = () => {
        const picLen = this.props.views.length;
        const h = picLen / 2;
        let values = [], otherHalf = [];

        for (let i = Math.floor(picLen / 2); i > 0; i--) {
            values.push({angle: this.MAX_ANGLE * (i / h), marginTop: this.V_OFFSET * i});
        }
        otherHalf = [...values].reverse();
        values = values.map(v => {
            return {angle: v.angle * -1, marginTop: v.marginTop}
        });
        if (picLen % 2 !== 0)
            values.push({angle: 0, marginTop: this.V_OFFSET / 3});
        const final = [...values, ...otherHalf];
        console.log(final);
        return final
    };

    _generateViews = () => {
        const views = [];

        this.props.views.forEach((view, index) => {
            views.push(React.cloneElement(view, {
                key: "" + index, style: {
                    ...view.props.style,
                    zIndex: 100 + index,
                    marginLeft: index === 0 ? 0 : -this.H_OFFSET,
                    marginTop: this.anglesAndPositions[index].marginTop,
                    transform: [{rotate: this.anglesAndPositions[index].angle + "deg" }]
                }
            }))
        });
        return views;
    };

    render() {
        return (
            <View style={[styles.container, this.props.containerStyle]}>
                {this._generateViews()}
            </View>
        );
    }
}

ViewsFan.defaultProps = {
    verticalOffset: 20,
    horizontalOffset: 20,
    reverseOverlap: false,
    maxAngle: 45
};

const styles = StyleSheet.create({
    container: {
        flexDirection: "row",
    }
});
