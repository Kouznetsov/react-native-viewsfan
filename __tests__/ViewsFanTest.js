import React from 'react';
import {View} from 'react-native';
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ViewsFan from "../ViewsFan"

configure({adapter: new Adapter()});

let array = [];

beforeAll(() => {
    const val = 5;
    for (let i = 0; i < val; i++)
        array.push(<View style={{borderRadius: 10, width: 70, height: 70, backgroundColor: i % 2 === 0 ? i === 0 ? "green" : "blue" : "red"}} />)
});

describe("ViewsFan class tests", () => {

    describe("props tests", () => {

        it("does not crash if no prop is passed", () => {
            const wrapper = shallow(<ViewsFan />);

            expect(wrapper).toMatchSnapshot();
        });

        it("does not crash if null is passed as views", () => {
            const wrapper = shallow(<ViewsFan views={null} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("does not crash if null is passed as verticalOffset", () => {
            const wrapper = shallow(<ViewsFan verticalOffset={null} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("does not crash if null is passed as horizontalOffset", () => {
            const wrapper = shallow(<ViewsFan horizontalOffset={null} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("does not crash if null is passed as maxAngle", () => {
            const wrapper = shallow(<ViewsFan maxAngle={null} />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("display tests", () => {

        it("displays without crashing", () => {
            const wrapper = shallow(<ViewsFan views={array} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("displays the same number of views that ara passed as parameter", () => {
            const wrapper = shallow(<ViewsFan views={array} />);

            expect(wrapper.children()).toHaveLength(array.length);
        });
    });
});