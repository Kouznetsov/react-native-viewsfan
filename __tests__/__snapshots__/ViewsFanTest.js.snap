// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewsFan class tests display tests displays without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <View
    key="0"
    style={
      Object {
        "backgroundColor": "green",
        "borderRadius": 10,
        "height": 70,
        "marginLeft": 0,
        "marginTop": 40,
        "transform": Array [
          Object {
            "rotate": "-36deg",
          },
        ],
        "width": 70,
        "zIndex": 100,
      }
    }
  />
  <View
    key="1"
    style={
      Object {
        "backgroundColor": "red",
        "borderRadius": 10,
        "height": 70,
        "marginLeft": -20,
        "marginTop": 20,
        "transform": Array [
          Object {
            "rotate": "-18deg",
          },
        ],
        "width": 70,
        "zIndex": 101,
      }
    }
  />
  <View
    key="2"
    style={
      Object {
        "backgroundColor": "blue",
        "borderRadius": 10,
        "height": 70,
        "marginLeft": -20,
        "marginTop": 6.666666666666667,
        "transform": Array [
          Object {
            "rotate": "0deg",
          },
        ],
        "width": 70,
        "zIndex": 102,
      }
    }
  />
  <View
    key="3"
    style={
      Object {
        "backgroundColor": "red",
        "borderRadius": 10,
        "height": 70,
        "marginLeft": -20,
        "marginTop": 20,
        "transform": Array [
          Object {
            "rotate": "18deg",
          },
        ],
        "width": 70,
        "zIndex": 103,
      }
    }
  />
  <View
    key="4"
    style={
      Object {
        "backgroundColor": "blue",
        "borderRadius": 10,
        "height": 70,
        "marginLeft": -20,
        "marginTop": 40,
        "transform": Array [
          Object {
            "rotate": "36deg",
          },
        ],
        "width": 70,
        "zIndex": 104,
      }
    }
  />
</View>
`;

exports[`ViewsFan class tests props tests does not crash if no prop is passed 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
/>
`;

exports[`ViewsFan class tests props tests does not crash if null is passed as horizontalOffset 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
/>
`;

exports[`ViewsFan class tests props tests does not crash if null is passed as maxAngle 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
/>
`;

exports[`ViewsFan class tests props tests does not crash if null is passed as verticalOffset 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
/>
`;

exports[`ViewsFan class tests props tests does not crash if null is passed as views 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
/>
`;
